### PR TITLE
plasma: 5.11.5 -> 5.12.1

### PR DESCRIPTION
**Pull request by adisbladis**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Tests are passing but have not had the time to actually try things out.

cc @ttuegel @peterhoeg @bkchr

### CODE REVIEW

File: pkgs/desktops/plasma-5/kde-gtk-config/default.nix [hunk 1]
```diff
@@ -2,19 +2,20 @@
...
+  knewstuff, gsettings_desktop_schemas
 }:
 
 mkDerivation {
   name = "kde-gtk-config";
   nativeBuildInputs = [ extra-cmake-modules ];
   buildInputs = [
     ki18n kio glib gtk2 gtk3 karchive kcmutils kconfigwidgets kiconthemes
-    knewstuff
+    knewstuff gsettings_desktop_schemas gsettings_desktop_schemas
```
Comment (bkchr) [inline pkgs/desktops/plasma-5/kde-gtk-config/default.nix]: Why did you add it twice?

File: pkgs/desktops/plasma-5/kwin/default.nix [hunk 2]
```diff
@@ -21,24 +21,18 @@ mkDerivation {
...
-        name = "cmake-3.10.diff";
-        # included upstream for kwin >= 5.11.95
-        url = "https://github.com/KDE/kwin/commit/cd544890ced4192.diff";
-        sha256 = "0z5nbcg712v10mskb7r9v0jcx5h8q4ixb7fjbb0kicmzsc266yd5";
-    })]
-    ;
+  patches = copyPathsToStore (lib.readPathsFromFile ./. ./series);
   CXXFLAGS = [
     ''-DNIXPKGS_XWAYLAND=\"${lib.getBin xwayland}/bin/Xwayland\"''
+    "-I${lib.getDev kwayland}/include/KF5"
```
Comment (bkchr) [inline pkgs/desktops/plasma-5/kwin/default.nix]: I created a patch upstream for this: https://phabricator.kde.org/D10373

**Review by bkchr — COMMENTED** [review]
Nice. I also started with this yesterday, but you was faster! :) I added two comments. I will compile it later and will report back.

File: pkgs/desktops/plasma-5/kwin/default.nix (hunk 2 above)
Comment (adisbladis) [reply pkgs/desktops/plasma-5/kwin/default.nix]: This is a much better fix. Changed to your patch :)

File: pkgs/desktops/plasma-5/kde-gtk-config/default.nix (hunk 1 above)
Comment (adisbladis) [reply pkgs/desktops/plasma-5/kde-gtk-config/default.nix]: Just a mistake. Fixed!

File: pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch [hunk 3]
```diff
@@ -814,32 +809,26 @@ index 9f875110..2a7a2a70 100644
...
+-  configDir=$XDG_CONFIG_HOME;
 -else
--  qdbus=qdbus
+-  configDir=${HOME}/.config; #this is the default, http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html
 +# we have to unset this for Darwin since it will screw up KDE's dynamic-loading
 +unset DYLD_FORCE_FLAT_NAMESPACE
 +
 +export XDG_CONFIG_HOME="${XDG_CONFIG_HOME:-$HOME/.config}"
-+@NIXPKGS_MKDIR@ -p "$XDG_CONFIG_HOME"
++_MKDIR@ -p "$XDG_CONFIG_HOME"
```
Comment (bkchr) [inline pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch]: `_MKDIR` needs to be `NIXPKGS_MKDIR`

File: pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch [hunk 4]
```diff
@@ -891,10 +875,12 @@ index 417a87d4..3f62745a 100644
 +gtk-button-images=1
 +EOF
  fi
+ sysConfigDirs=${XDG_CONFIG_DIRS:-/etc/xdg}
  
+-# We need to create config folder so we can write startupconfigkeys
 -mkdir -p $configDir
 +# Set the default GTK 3 theme
-+gtk3_settings="$XDG_CONFIG_HOME/gtk-3.0/settings.ini"
++_settings="$XDG_CONFIG_HOME/gtk-3.0/settings.ini"
```
Comment (bkchr) [inline pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch]: Also looks not good.

**Comment by bkchr** [comment]
@adisbladis, I applied the following changes:
```diff --git a/pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch b/pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch
index f3aa4743bb2..e012537e402 100644
--- a/pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch
+++ b/pkgs/desktops/plasma-5/plasma-workspace/plasma-workspace.patch
@@ -828,7 +828,7 @@ index 8ac47aa..49970ef 100644
 +unset DYLD_FORCE_FLAT_NAMESPACE
 +
 +export XDG_CONFIG_HOME="${XDG_CONFIG_HOME:-$HOME/.config}"
-+_MKDIR@ -p "$XDG_CONFIG_HOME"
++@NIXPKGS_MKDIR@ -p "$XDG_CONFIG_HOME"
 +
 +# The KDE icon cache is supposed to update itself
 +# automatically, but it uses the timestamp on the icon
@@ -880,7 +880,7 @@ index 8ac47aa..49970ef 100644
 -# We need to create config folder so we can write startupconfigkeys
 -mkdir -p $configDir
 +# Set the default GTK 3 theme
-+_settings="$XDG_CONFIG_HOME/gtk-3.0/settings.ini"
++gtk3_settings="$XDG_CONFIG_HOME/gtk-3.0/settings.ini"
 +breeze_gtk3="/run/current-system/sw/share/themes/Breeze/gtk-3.0"
 +if ! [ -e "$gtk3_settings" ] && [ -e "$breeze_gtk" ]; then
 +    mkdir -p $(dirname "$gtk3_settings")
@@ -1033,7 +1033,7 @@ index 8ac47aa..49970ef 100644
 -        XCURSOR_THEME="$kcminputrc_mouse_cursortheme"
 -        export XCURSOR_THEME
 +if [ -n "$kcminputrc_mouse_cursortheme" -o -n "$kcminputrc_mouse_cursorsize" ]; then
-+    kapplymousetheme "$kcminputrc_mouse_cursortheme" "$kcminputrc_mouse_cursorsize"
++    #kapplymousetheme "$kcminputrc_mouse_cursortheme" "$kcminputrc_mouse_cursorsize"
 +    if [ $? -eq 10 ]; then
 +        export XCURSOR_THEME=breeze_cursors
 +    elif [ -n "$kcminputrc_mouse_cursortheme" ]; then
@@ -1106,7 +1106,7 @@ index 8ac47aa..49970ef 100644
 -
  # Make sure that D-Bus is running
 -if qdbus >/dev/null 2>/dev/null; then
-+if ! @NIXPKGS_QDBUS@ >/dev/null 2>/dev/null; then
++if @NIXPKGS_QDBUS@ >/dev/null 2>/dev/null; then
      : # ok
  else
      echo 'startplasmacompositor: Could not start D-Bus. Can you call qdbus?'  1>&2
```

After applying these changes, I can start the wayland session :)

Besides that, everything seems to be working for me.

**Comment by coreyoconnor** [comment]
xf86 input libinput is now required by plasma-desktop. Otherwise the "mouse" system settings is not present:

* https://github.com/coreyoconnor/nixpkgs/commit/d8689865bf8af6159333cba1e2f953f3eec27720

**Comment by adisbladis** [comment]
@bkchr Thanks! Applied and rebased.

@coreyoconnor 
> xf86 input libinput is now required by plasma-desktop. Otherwise the "mouse" system settings is not present
Perfect! Applied your patch.

**Review by coreyoconnor — APPROVED** [review]

**Comment by adisbladis** [comment]
Now I have actually been running on this for a few hours and things seem to work as expected! :)

File: pkgs/desktops/plasma-5/kwin/default.nix [hunk 5]
```diff
@@ -21,22 +21,20 @@ mkDerivation {
...
-    ++ [(fetchpatch {
-        name = "cmake-3.10.diff";
-        # included upstream for kwin >= 5.11.95
-        url = "https://github.com/KDE/kwin/commit/cd544890ced4192.diff";
-        sha256 = "0z5nbcg712v10mskb7r9v0jcx5h8q4ixb7fjbb0kicmzsc266yd5";
-    })]
-    ;
+  patches = copyPathsToStore (lib.readPathsFromFile ./. ./series) ++ [
+    (fetchpatch {
+      url = "https://phabricator.kde.org/file/data/ojbvgljuhqxztrdcicoy/PHID-FILE-6wvjkli2blrgw7jei3qg/D26720.diff";
```
Comment (bkchr) [inline pkgs/desktops/plasma-5/kwin/default.nix]: The url does not work for me.
The following works for me:
https://phabricator.kde.org/file/data/5bu6h2c2ksaqctyy275b/PHID-FILE-jtbgqz2heluxsovlqyv3/D10373.diff

Suspicious, that your url also has a different filename?

**Comment by etu** [comment]
This also just became security related with [CVE-2018-6791](https://www.kde.org/info/security/advisory-20180208-2.txt).

**Comment by ttuegel** [comment]
We should implement this; I think a few users are testing Plasma on Wayland, although we don't have it enabled by default yet: https://mail.kde.org/pipermail/distributions/2017-December/000260.html

File: pkgs/desktops/plasma-5/kwin/default.nix (hunk 5 above)
Comment (adisbladis) [reply pkgs/desktops/plasma-5/kwin/default.nix]: Strange. That was working at some point.

How stable are phabricator urls?

Comment (bkchr) [reply pkgs/desktops/plasma-5/kwin/default.nix]: I don't know :(

Comment (bkchr) [reply pkgs/desktops/plasma-5/kwin/default.nix]: @adisbladis if you want to be on the save side, the patch got merged: https://github.com/KDE/kwin/commit/6e5f5d92daab4c60f7bf241d90a91b3bea27acfd

Comment (adisbladis) [reply pkgs/desktops/plasma-5/kwin/default.nix]: Perfect! I'll use the github link.

**Comment by adisbladis** [comment]
@ttuegel Implemented! https://github.com/adisbladis/nixpkgs/blob/5708382530a6056a63cbcbc6429cc6cd2ff4fca8/nixos/modules/services/x11/desktop-managers/plasma5.nix#L69

File: pkgs/desktops/plasma-5/kwin/default.nix [hunk 6]
```diff
@@ -21,22 +21,21 @@ mkDerivation {
...
   outputs = [ "bin" "dev" "out" ];
-  patches = copyPathsToStore (lib.readPathsFromFile ./. ./series)
-    ++ [(fetchpatch {
-        name = "cmake-3.10.diff";
-        # included upstream for kwin >= 5.11.95
-        url = "https://github.com/KDE/kwin/commit/cd544890ced4192.diff";
-        sha256 = "0z5nbcg712v10mskb7r9v0jcx5h8q4ixb7fjbb0kicmzsc266yd5";
-    })]
-    ;
+  patches = copyPathsToStore (lib.readPathsFromFile ./. ./series) ++ [
```
Comment (peterhoeg) [inline pkgs/desktops/plasma-5/kwin/default.nix]: Please add in a comment mentioning when we can drop this patch again.

Comment (adisbladis) [reply pkgs/desktops/plasma-5/kwin/default.nix]: Done

**Comment by adisbladis** [comment]
Plasma `5.12.1` was released. Bumped PR and rebased on current master.

**Comment by peterhoeg** [comment]
Do you mind testing this together with #34936?

**Comment by adisbladis** [comment]
@peterhoeg My desktop has already been running on both of these PRs for a day :)

**Comment by peterhoeg** [comment]
Then LGTM!

**Comment by bkchr** [comment]
I also have Plasma 5.12 and frameworks 5.43 running on my Laptop and everything seems to be fine :)

**Pull request merged**